### DISCUSSION TITLE
[@types/react-redux] null mapDispatchtoProps doesn't provide correct parameter.

### DISCUSSION THREAD

**Discussion by saraid** [DefinitelyTyped/DefinitelyTyped#59297] [Issues with a @types package] [answered]
Let's say I define no mapStateToProps or mapDispatchToProps. Instead, I define a mergeProps like so:

```ts
const mergeProps(
  stateProps: undefined,
  { dispatch }: { dispatch: Dispatcher },
  externalProps: ExternalProps
): Props => {
  return {
    whatever: () => dispatch(someAction())
  }
}

export default connect(undefined, undefined, mergeProps)(MyComponent);
```

This is functional code, because the following passes `{ dispatch }` into the mergeProps function:
https://github.com/reduxjs/react-redux/blob/4d384029e3fa0355e762f40409a5dac4358a3de3/src/connect/mapDispatchToProps.ts#L14-L20

However, the type definition expects this to be `undefined` instead:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/998fe1077af548a7c97fcee5f2057bdb04d3855c/types/react-redux/index.d.ts#L233-L243

This could be fixed by doing this instead:
```ts
    <no_state = {}, no_dispatch = {}, TOwnProps = {}, TMergedProps = {}>(
        mapStateToProps: null | undefined,
        mapDispatchToProps: null | undefined,
        mergeProps: MergeProps<undefined, { dispatch: Dispatcher }, TOwnProps, TMergedProps>,
    ): InferableComponentEnhancerWithProps<TMergedProps, TOwnProps>;


    <TStateProps = {}, no_dispatch = {}, TOwnProps = {}, TMergedProps = {}, State = DefaultState>(
        mapStateToProps: MapStateToPropsParam<TStateProps, TOwnProps, State>,
        mapDispatchToProps: null | undefined,
        mergeProps: MergeProps<TStateProps, { dispatch: Dispatcher }, TOwnProps, TMergedProps>,
    ): InferableComponentEnhancerWithProps<TMergedProps, TOwnProps>;
```

Though I suppose we'd want to name it something appropriate.

Does this seem correct to others? I've never contributed to DT before, so I'm unsure about process; should I open a PR?

[a comment by an automated account is not shown]

**Comment by markerikson** [comment, marked as answer]
That seems plausible, yes.  (Although as a maintainer I'd really question why anyone would write their component that way to begin with :)  and if you are actually doing that, I'm genuinely curious to hear why.)

And yeah, I think you can just file a PR to update the types here in DT.  If you do, could you also file an issue on the React-Redux repo as well? We've migrated React-Redux itself to TS in v8, based on these types.

**Reply by saraid** [reply]
Alas, I didn't write it. I'm just maintaining it right now. 😛 I'm currently in the process of turning on `strictFunctionTypes` for a bunch of legacy code and this is something that got revealed by that. Rewriting it to add in a `mapDispatchToProps = ({ dispatch }) => ({ dispatch })` should also probably address the issue, but I feel that I shouldn't have to add a new function and change working functionality to get the types right.
